zilencer/views.py

```diff
@@ -216,13 +220,73 @@ def unregister_all_remote_push_devices(
     return json_success(request)
 
 
+def delete_duplicate_registrations(
+    registrations: List[RemotePushDeviceToken], server_id: int, user_id: int, user_uuid: str
+) -> List[RemotePushDeviceToken]:
+    """
+    When migrating to support registration by UUID, we introduced a bug where duplicate
+    registrations for the same device+user could be created - one by user_id and one by
+    user_uuid. Given no good way of detecting these duplicates at database level, we need to
+    take advantage of the fact that when a remote server sends a push notification request
+    to us, it sends both user_id and user_uuid of the user.
+
+    This function, knowing the user_id and user_uuid of the user, can detect duplicates
+    and delete the legacy user_id registration if appropriate.
+
+    Return the list of registrations with the user_id-based duplicates removed.
+    """
+
+    # All registrations passed here should be of the same kind (apple vs android).
+    assert len({registration.kind for registration in registrations}) == 1
+    kind = registrations[0].kind
+
+    tokens = [device.token for device in registrations]
+    tokens_counter = Counter(tokens)
+
+    tokens_to_deduplicate = []
+    for key in tokens_counter:
+        if tokens_counter[key] <= 1:
+            continue
+        if tokens_counter[key] > 2:
+            raise AssertionError(
+                f"More than two registrations for token {key} for user id:{user_id} uuid:{user_uuid}, shouldn't be possible"
+            )
+        assert tokens_counter[key] == 2
+        tokens_to_deduplicate.append(key)
+
+    if not tokens_to_deduplicate:
+        return registrations
+
+    logger.info(
+        "Deduplicating push registrations for server id:%s user id:%s uuid:%s and tokens:%s",
+        server_id,
+        user_id,
+        user_uuid,
+        sorted(tokens_to_deduplicate),
+    )
+    RemotePushDeviceToken.objects.filter(
+        token__in=tokens_to_deduplicate, kind=kind, server_id=server_id, user_id=user_id
+    ).delete()
+
+    deduplicated_registrations_to_return = []
+    for registration in registrations:
+        if registration.token in tokens_to_deduplicate and registration.user_id is not None:
+            # user_id registrations are the ones we deleted
+            continue
```

Comment:
Is the existing invariant that every record for a given (server_id, kind, token) triple will have either a non-null user_id, or a non-null user_uuid, but not both?

If we were to get here and somehow have two records with the same kind, server_id, and token and *both* have the same non-null user_id — perhaps one of them has a user_uuid and the other doesn't — then I think this code would end up deleting both of them. If we have a database constraint that prevents that, then great. If not, then probably it still can't happen, but it'd be nice to have some kind of assert in this code so that we don't take things further awry if it somehow does happen.

Reply:
Correct, that's the existing invariant.